src/API.jl

```diff
@@ -892,75 +892,94 @@ function build(ctx::Context, pkgs::Vector{PackageSpec}; verbose=false, kwargs...
+        pkg = Base.PkgId(first(dep), last(dep).name)
+        Base.in_sysimage(pkg) && return nothing
+        deps = [Base.PkgId(last(x), first(x)) for x in last(dep).deps]
+        return pkg => filter!(!Base.in_sysimage, deps)
+    end
+    depsmap = Dict(Iterators.filter(!isnothing, deps_pair_or_nothing)) #flat map of each dep and its deps
     
     if ctx.env.pkg !== nothing && isfile( joinpath( dirname(ctx.env.project_file), "src", ctx.env.pkg.name * ".jl") )
-        push!(pkgids, Base.PkgId(ctx.env.pkg.uuid, ctx.env.pkg.name))
-        push!(pkg_dep_uuid_lists, collect(keys(ctx.env.project.deps)))
+        depsmap[Base.PkgId(ctx.env.pkg.uuid, ctx.env.pkg.name)] = [
+            Base.PkgId(last(x), first(x)) 
+            for x in ctx.env.project.deps if !Base.in_sysimage(Base.PkgId(last(x), first(x)))
+        ]
     end    
     
-    was_processed = Dict{Base.UUID,Base.Event}()
-    was_recompiled = Dict{Base.UUID,Bool}()
-    for pkgid in pkgids
-        was_processed[pkgid.uuid] = Base.Event()
-        was_recompiled[pkgid.uuid] = false
-    end
-    
-    function is_stale(paths, sourcepath)
-        for path_to_try in paths::Vector{String}
-            staledeps = Base.stale_cachefile(sourcepath, path_to_try, Base.TOMLCache())
-            staledeps === true && continue
-            return false
-        end
-        return true
+    was_processed = Dict{Base.PkgId,Base.Event}()
+    was_recompiled = Dict{Base.PkgId,Bool}()
+    for pkgid in keys(depsmap)
+        was_processed[pkgid] = Base.Event()
+        was_recompiled[pkgid] = false
     end
     
     errored = false
-    @sync for (i, pkg) in pairs(pkgids)
+    toml_c = Base.TOMLCache()
+    @sync for (pkg, deps) in depsmap
         paths = Base.find_all_in_cache_path(pkg)
-        sourcepath = Base.locate_package(pkg)
+        sourcepath = Base.locate_package(pkg, toml_c)
         sourcepath === nothing && continue
         # Heuristic for when precompilation is disabled
-        occursin(r"\b__precompile__\(\s*false\s*\)", read(sourcepath, String)) && continue
+        if occursin(r"\b__precompile__\(\s*false\s*\)", read(sourcepath, String))
+            notify(was_processed[pkg])
+            continue
+        end
         
         @async begin
-            for dep_uuid in pkg_dep_uuid_lists[i] # wait for deps to finish
-                wait(was_processed[dep_uuid])
+            for dep in deps # wait for deps to finish
+                wait(was_processed[dep])
             end
             
             # skip stale checking and force compilation if any dep was recompiled in this session
-            any_dep_recompiled = any(map(dep_uuid->was_recompiled[dep_uuid], pkg_dep_uuid_lists[i]))
-            if !errored && (any_dep_recompiled || is_stale(paths, sourcepath))
+            any_dep_recompiled = any(map(dep->was_recompiled[dep], deps))
+            if !errored && (any_dep_recompiled || _is_stale(paths, sourcepath, toml_c))
```

Comment:
If we are going to rely on `Base.TOMLCache` to be "`async`-safe" (which is, BTW, pretty easy to break; e.g., a `@debug` can ruin this property), I don't think it's necessary to move `_is_stale` out. It's actually where closure is useful. Although there is nothing bad in factoring out `_is_stale` this way.

Reply:
@staticfloat did some probing of the list of toml files that were accessed when no tomlcaches were provided, and it was just the env manifest and project files, so the thinking was that it was async safe as it's read-only after the first use in the for-loop. Would that be broken by `@debug`?

Reply:
It'd be good to resolve this, and merge, to fix #2035

@fredrikekre @KristofferC

Reply:
`parsed_toml(cache::TOMLCache, project_file::String)` reuses `TOML.Parser`: https://github.com/JuliaLang/julia/blob/2a36f83ce9bdfd17c572472f60b830aedcf181f0/base/loading.jl#L174-L175

So, for example, `TOML.parse(::TOML.Parser)` has to be async-safe. It probably is and likely be so forever but relying on "likely" is not the best approach especially in concurrent programming. For example, maybe somebody will implement `TOML.parse(::IO)` at some point and the parsing can happen in a streaming fashion.

(It's also a bit concerning that the I/O `read(project_file, String)` is happening in `get!` but that's probably handled by the `.age` field in `Dict`.)

Anyway, I don't think my comment should block this PR. We can fix it in `Base` if we want.

Reply:
Ok, merging this as-is seems like the thing to do for now